common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/OAuth2Strategy.java

```diff
@@ -154,6 +156,12 @@ protected Uri createAuthorizationUri() {
      */
     protected abstract void validateTokenRequest(GenericTokenRequest request);
 
+    /**
+     * Abstract method for validating the http response from server for token request.
+     * @param response
+     */
+    protected abstract void validateTokenResponse(final HttpResponse response) throws ServiceException;
```

Comment:
Thank you for the comment @iambmelt I don't think we should use the generics of token response here. As the validation should happen before the initialization of token response object. The return of the OAuth2 post is an HTTPResponse type. ( [RFC ](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-protocols-oauth-code#use-the-authorization-code-to-request-an-access-token)) The aim of `validateTokenResponse` is to validate if the http response could be in well-format of JSON so that it could be able to init the GenericTokenResponse object. Thus, the input parameters here should be  HTTPResponse instead of GenericTokenResponse. 

/cc @addev-ashish @shoatman 

Please let me know if I misunderstood the requirement of the validation functions here.

Reply:
`These method signatures may need to get updated with thrown Exception types depending on the nature of their validation. Most likely ServiceException would be the thrown error.` 👍

Reply:
`Gson` will tell us whether the JSON is malformed or not. We shouldn't have to verify this 'by-hand'.

The way this should work is that `Gson` will parse the `HttpResponse` body into a defined POJO (such as `MicrosoftStsTokenResponse`, `ActiveDirectoryFederationServicesTokenResponse`,  `AzureActiveDirectoryB2CTokenResponse`, etc) and _subclasses_ of `OAuth2Strategy` will override `validateTokenResponse(GenericTokenResponse)` with an implementation that enforces whatever requirements are necessary to make sure that the result can be processed on the client side ('processed' meaning written to cache, broker, etc).

For example:
```java
public abstract class OAuth2Strategy
        <GenericAccessToken extends AccessToken,
                GenericAccount extends Account,
                GenericAuthorizationRequest extends AuthorizationRequest,
                GenericAuthorizationResponse extends AuthorizationResponse,
                GenericAuthorizationStrategy extends AuthorizationStrategy,
                GenericOAuth2Configuration extends OAuth2Configuration,
                GenericRefreshToken extends RefreshToken,
                GenericTokenRequest extends TokenRequest,
                GenericTokenResponse extends TokenResponse,
                GenericTokenResult extends TokenResult> {

    protected void validateTokenResponse(GenericTokenResponse response) {
        // Validate that the POJO created by Gson contains non-null values for
        // expires_in, access_token, token_type, refresh_token, id_token, etc
    }

}
```

And subclasses would `@Override` this method validating their own needs, for example...
```java
public class MicrosoftStsOAuth2Strategy
        extends OAuth2Strategy
        <MicrosoftStsAccessToken,
                MicrosoftStsAccount,
                MicrosoftStsAuthorizationRequest,
                AuthorizationResponse,
                AuthorizationStrategy,
                MicrosoftStsOAuth2Configuration,
                MicrosoftStsRefreshToken,
                TokenRequest,
                MicrosoftStsTokenResponse,
                TokenResult> {

    protected void validateTokenResponse(final MicrosoftStsTokenResponse response) {
        // First call the super validator
        super(response);
        // Validate that the POJO created by Gson contains non-null values for
        // scope, client_info, and other v2 specific things
    }

}
```

So what we're really doing here is trusting `Gson` to do its job and serialize the JSON response into the POJOs, throwing a `JsonSyntaxException` if there are problems and then we validate the outputs contain everything we need.